source/blog/2021-03-29-dbal-2.13.md

```diff
@@ -0,0 +1,142 @@
+---
+title: "New Release: Doctrine DBAL 2.13 with Deprecations and Forward Compatibility"
+authorName: Benjamin Eberlei
+authorEmail: kontakt@beberlei.de
+permalink: /2021/03/29/dbal-2.13.html
+---
+
+We have released DBAL 2.13, what we plan to be the last minor version in the 2.x family of Doctrine DBAL.
+This release includes additional forward compatibilty to DBAL 3 around
+`Statement` and `Result` API and with an integration in our new
+deprecations logging library.
+
+In addition this DBAL release re-enables PHP 7.1 and 7.2 compatibility to give
+as much flexibility as possible to everyone with forward compatibility.
+
+### Statement and Result Forward Compatibility
+
+DBAL 3.0 extracts all fetch-methods from the `Statement` class and moved them
+to a new `Result` class that is returned from `Statement::execute`. We have
+backported this API to 2.13 - so that you can support writing code for both
+DBAL 2 and 3 at the same time.
+
+Old code:
+
+```php
+$statement = $connection->prepare('SELECT * FROM tbl WHERE col = ?');
+$statement->bindParam(1, $value);
+$statement->execute();
+
+while ($row = $statement->fetch()) {
+}
+
+$connection->executeQuery('SELECT * FROM tbl')->fetchAll();
+```
+
+New Code:
+
+```php
+$statement = $connection->prepare('SELECT * FROM tbl WHERE col = ?');
+$statement->bindParam(1, $value);
+$result = $statement->execute();
+
+while ($row = $result->fetchAssociative()) {
+}
+
+$connection->executeStatement('SELECT * FROM tbl')->fetchAllAssociative();
+```
+
+
+The Result Fetching API was improved to use more human-readable names:
+
+```php
+// Old
+$stmt->fetch();
+$stmt->fetch(PDO::FETCH_ARRAY);
+$stmt->fetchColumn();
+$stmt->fetchAll();
+
+// New
+$result = $stmt->execute();
+$result->fetchAssociative();
+$result->fetchNumeric();
+$result->fetchOne();
+$result->fetchAllAssociative();
+```
+
+Many more changes have been made on the public API and also for the internals,
+but these are the most common ones.
+
+Thank you to [mdumoulin](https://github.com/mdumoulin) for the work on
+improving the forward compatiblity.
+
+### Deprecations Logging
+
+We have wrestled internally for a long time with the strategy on runtime
+deprecation going forward and settled on introducing a tiny abstraction on
+top of `@trigger_error()`.
```

Comment:
I think the whole point of this abstraction is that it doesn't unconditionally call `@trigger_error()`. I'd rephrase it to "an API for logging the usage of deprecated APIs". Something like that.